File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/BlockState.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface BlockStateLib: Library {
    fun BlockState_destroy(handle: Pointer)
    fun BlockState_create(name: Slice): Pointer
    fun BlockState_with_property(handle: Pointer, key: Slice, value: Slice): ResultPointerInt
    fun BlockState_name(handle: Pointer, write: Pointer): Unit
    fun BlockState_properties_json(handle: Pointer, write: Pointer): Unit
}
/** A block state: a block name plus its properties. Port of the old
*`BlockStateWrapper` / `blockstate_*` fns.
*/
class BlockState internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class BlockStateCleaner(val handle: Pointer, val lib: BlockStateLib) : Runnable {
        override fun run() {
            lib.BlockState_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, BlockState.BlockStateCleaner(handle, BlockState.lib));
    }

    companion object {
        internal val libClass: Class<BlockStateLib> = BlockStateLib::class.java
        internal val lib: BlockStateLib = Native.load("nucleation", libClass)
        @JvmStatic
        
        fun create(name: String): BlockState {
            val nameSliceMemory = PrimitiveArrayTools.borrowUtf8(name)
            
            val returnVal = lib.BlockState_create(nameSliceMemory.slice);
            try {
                val selfEdges: List<Any> = listOf()
                val handle = returnVal 
                val returnOpaque = BlockState(handle, selfEdges, true)
                return returnOpaque
            } finally {
                nameSliceMemory.close()
            }
        }
    }
    
    /** A copy of this block state with `key=value` set; the original is
    *unchanged.
    */
    fun withProperty(key: String, value: String): Result<BlockState> {
        val keySliceMemory = PrimitiveArrayTools.borrowUtf8(key)
        val valueSliceMemory = PrimitiveArrayTools.borrowUtf8(value)
        
        val returnVal = lib.BlockState_with_property(handle, keySliceMemory.slice, valueSliceMemory.slice);
        try {
            val nativeOkVal = returnVal.getNativeOk();
            if (nativeOkVal != null) {
                val selfEdges: List<Any> = listOf()
                val handle = nativeOkVal 
                val returnOpaque = BlockState(handle, selfEdges, true)
                return returnOpaque.ok()
            } else {
                return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
            }
        } finally {
            keySliceMemory.close()
            valueSliceMemory.close()
        }
    }
    
    fun name(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.BlockState_name(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }
    
    /** The properties as a JSON object of string→string (the old
    *`CPropertyArray`).
    */
    fun propertiesJson(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.BlockState_properties_json(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }

}

File: bindings/php/BlockState.php
<?php
namespace Stencil;

final class BlockState {
    /** @internal */
    public \FFI\CData $ptr;
    private bool $owned;

    /** @internal */
    public function __construct(\FFI\CData $ptr, bool $owned) {
        $this->ptr = $ptr;
        $this->owned = $owned;
    }

    public static function create(string $name) {
        $__n0 = strlen($name);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $name, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $ret = Lib::ffi()->BlockState_create($__view0);
        return new BlockState($ret, true);
    }

    public function withProperty(string $key, string $value) {
        $__n0 = strlen($key);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $key, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $__n1 = strlen($value);
        $__view1 = Lib::ffi()->new('DiplomatStringView');
        if ($__n1 > 0) {
            $__buf1 = Lib::ffi()->new("uint8_t[" . $__n1 . "]", false);
            \FFI::memcpy($__buf1, $value, $__n1);
            $__view1->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf1[0]));
        } else {
            $__view1->data = null;
        }
        $__view1->len = $__n1;
        $result = Lib::ffi()->BlockState_with_property($this->ptr, $__view0, $__view1);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return new BlockState($result->ok, true);
    }

    public function name() {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        Lib::ffi()->BlockState_name($this->ptr, $write);
        return Lib::readAndFreeWrite($write);
    }

    public function propertiesJson() {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        Lib::ffi()->BlockState_properties_json($this->ptr, $write);
        return Lib::readAndFreeWrite($write);
    }

    public function __destruct() {
        if ($this->owned) {
            Lib::ffi()->BlockState_destroy($this->ptr);
        }
    }
}
